### PR TITLE
Fixed #221. The problem with complex sqrt

### LINKED ISSUE #221: `(1e10 - 1e-10i) sqrt` does not work
I try to compute: ``(1e10 - 1e-10i) sqrt`` and I got a division by zero error.

I'm wondering if this is not the same bug than in Math.js: https://github.com/josdejong/mathjs/pull/208
Any comments on that @nicolas-cellier-aka-nice ?

### PR DESCRIPTION
**Pull request by olekscode**

### CODE REVIEW

File: src/Math-Tests-Complex/PMComplexTest.class.st [hunk 1]
```diff
@@ -178,6 +178,36 @@ PMComplexTest >> testBug1 [
...
+
+{ #category : #tests }
+PMComplexTest >> testCloseToRealWithPrecision [
+
+	self assert: (2 + 0.001i closeTo: 2 precision: 0.01).
+	self assert: (2.001 + 0.001i closeTo: 2 precision: 0.01).
+]
+
+{ #category : #tests }
+PMComplexTest >> testCloseToWithPrecision [
```
Comment (hemalvarambhia) [inline src/Math-Tests-Complex/PMComplexTest.class.st:204]: Perhaps a test where the number is outside the precision required?

**Review by hemalvarambhia — APPROVED** [review]
I don't recall comparing complex numbers directly, only their moduli. Im

One small comment both otherwise, great job fixing a defect.

File: src/Math-Tests-Complex/PMComplexTest.class.st [hunk 2]
```diff
@@ -178,6 +178,36 @@ PMComplexTest >> testBug1 [
...
+{ #category : #tests }
+PMComplexTest >> testCloseTo [
+
+	self assert: 2 + 3.000000000000001i closeTo: 2 + 3i.
+	self assert: 2.000000000000001 + 3i closeTo: 2 + 3i.
+	self assert: 2.000000000000001 + 3.000000000000001i closeTo: 2 + 3i.
+]
+
+{ #category : #tests }
+PMComplexTest >> testCloseToReal [
```
Comment (hemalvarambhia) [inline src/Math-Tests-Complex/PMComplexTest.class.st:190]: What about a test where you compare a complex number with a real?

E.g. deny: 2 + 3 i closeTo: 2.000001

File: src/Math-Tests-Complex/PMComplexTest.class.st (hunk 1 above)
Comment (olekscode) [reply src/Math-Tests-Complex/PMComplexTest.class.st:207]: Done

File: src/Math-Tests-Complex/PMComplexTest.class.st (hunk 2 above)
Comment (olekscode) [reply src/Math-Tests-Complex/PMComplexTest.class.st:190]: Done

**Review by hemalvarambhia — APPROVED** [review]

**Comment by hemalvarambhia** [comment]
:shipit:

**Comment by nicolas-cellier-aka-nice** [comment]
Yes, I first thought using the same formulation which is a bit more robust.<br>
Reference "How to Find the Square Root of a Complex Number" by Stanley Rabinowitz 1993.

However, this implementation still suffers from potential floating point overflow/underflow.<br>
This is why I suggest something even more robust (and more accurate) like:

Implementing Complex Elementary Function Using Exception Handling<br>
ACM Transactions on Mathematical Software - October 1994<br>
Ping Tang and 3 other authors

See https://source.squeak.org/trunk/Kernel-nice.1459.diff

**Comment by olekscode** [comment]
Oh, this is cool!
Thanks @nicolas-cellier-aka-nice 

I like the idea of scaling

**Comment by olekscode** [comment]
I will merge this PR for now and open a new issue for a more robust sqrt implementation

**Comment by olekscode** [comment]
> I will merge this PR for now and open a new issue for a more robust sqrt implementation

Issue #257

**Pull request merged**